File: dotnet/src/MyDotey.QuantileEstimator/Gk/GkQuantileEstimator.cs
using System;
using System.Collections.Generic;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
namespace MyDotey.Quantile.Gk
{
    public class GkQuantileEstimator<T> : IQuantileEstimator<T>
    {
        private GkQuantileEstimatorConfig<T> _config;
        private QuantileEstimationGK<T> _quantileEstimationGK;

        public GkQuantileEstimator(GkQuantileEstimatorConfig<T> config)
        {
            if (config == null)
                throw new ArgumentNullException("config is null");

            _config = config;
            _quantileEstimationGK = new QuantileEstimationGK<T>(_config.Comparer, _config.Error,
                    _config.CompactThreshold);
        }

        public virtual void Add(T value)
        {
            _quantileEstimationGK.Insert(value);
        }

        public virtual Dictionary<Double, T> Get(List<Double> quantiles)
        {
            Dictionary<Double, T> results = null;
            foreach (Double quantile in quantiles)
            {
                T result = _quantileEstimationGK.Query(quantile);
                if (_quantileEstimationGK.IsEmpty)
                    return null;

                if (results == null)
                    results = new Dictionary<Double, T>();

                results[quantile] = result;
            }

            if (results == null)
                results = new Dictionary<Double, T>();

            return results;
        }
    }
}

File: java/quantile-estimator/src/main/java/org/mydotey/quantile/gk/GkQuantileEstimator.java
package org.mydotey.quantile.gk;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.mydotey.quantile.QuantileEstimator;

/**
 * @author koqizhao
 *
 * Apr 2, 2018
 */
public class GkQuantileEstimator<T> implements QuantileEstimator<T> {

    private GkQuantileEstimatorConfig<T> _config;
    private QuantileEstimationGK<T> _quantileEstimationGK;

    public GkQuantileEstimator(GkQuantileEstimatorConfig<T> config) {
        Objects.requireNonNull(config, "config is null");

        _config = config;
        _quantileEstimationGK = new QuantileEstimationGK<>(_config.getComparator(), _config.getError(),
                _config.getCompactThreshold());
    }

    @Override
    public void add(T value) {
        _quantileEstimationGK.insert(value);
    }

    @Override
    public Map<Double, T> get(List<Double> quantiles) {
        HashMap<Double, T> results = null;
        for (Double quantile : quantiles) {
            T result = _quantileEstimationGK.query(quantile);
            if (_quantileEstimationGK.isEmpty())
                return null;

            if (results == null)
                results = new HashMap<>();

            results.put(quantile, result);
        }

        if (results == null)
            results = new HashMap<>();

        return results;
    }

}
